### PR TITLE
Rust core

### PR DESCRIPTION
**Pull request by fjarri**
**Type of PR:**
- Feature

**Required reviews:** 
- 3

**What this does:**

This PR replaces the `core.py` and `versioning.py` with the Rust crate https://github.com/nucypher/nucypher-core and its Python bindings, to allow for easy interoperability with JS.

It is mostly a porting job, but there were some changes introduced in the protocol and elsewhere:
- `umbral` is imported as a submodule of `nucypher_core` instead of as a separate package (since objects from different extensions are not interoperable)
- Protocol objects use MessagePack for serialization (at least in the current protocol version)
- Protocol object constructors are now invoked just with the respective type object (that is, as `__init()__`) instead of a special classmethod. Namely `MessageKit.author()`, `EncryptedKeyFrag.author()`, `TreasureMap.construct_by_publisher()`, `EncryptedTreasureMap.construct_by_publisher()`, ` ReencryptionRequest.from_treasure_map()`, `ReencryptionResponse.construct_by_ursula()`, `RevocationOrder.author()`, `NodeMetadata.author()` were removed.
- Wherever Ethereum addresses are present in the API, they are considered to be canonical addresses (a special `Address` type in Rust, or just length-20 `bytes` on the Python side)
- `AuthorizedKeyFrag` is not public anymore; `EncryptedKeyFrag.decrypt()` now decrypts into `AuthorizedKeyFrag` (internally) and verifies that producing a `VerifiedKeyFrag`
- Similarly with treasure maps, `AuthorizedTreasureMap` is not public, and `EncryptedTreasureMap.decrypt()` produces `TreasureMap` directly.
- `ReencryptionRequest` does not take a treasure map in its constructor, but instead the `EncryptedKeyFrag` directly (to avoid errors associated with the lookup in the destinations map)
- `FleetStateChecksum` was moved to `nucypher-core` from `perception.py`
- `MetadataRequest` takes a `FleetStateChecksum` object instead of a string
- `MetadataRequest` constructor does not take `None` as `announce_nodes` anymore; if there are no nodes to announce, give it an empty list
- `MetadataResponsePayload` constructor does not take `this_node` separately (merge it with other nodes), and `other_nodes` was renamed to `announce_nodes` (does not take `None`; pass an empty list)

### CODE REVIEW

**Comment by piotr-roslaniec** [comment]
`newsfragments/2832.feature.rst` is an empty file.

**Comment by KPrasch** [comment]
> `newsfragments/2832.feature.rst` is an empty file.

IMO It's generally acceptable that a PR has an empty newsfragment file.  It will result in a link to the PR being placed in the release notes without an annotation.  No harm in dropping some kind of note here though.

**Comment by vepkenez** [comment]
Maybe I can try getting those docker based tests working

**Comment by KPrasch** [comment]
> Maybe I can try getting those docker based tests working

Sounds great.  Note that this PR is based on #2835 which somehow passes! 🤔

**Comment by derekpierre** [comment]
> newsfragments/2832.feature.rst is an empty file.

I think such an impactful change deserves a note.

**Review by piotr-roslaniec — APPROVED** [review]
LGTM :tada:

File: deploy/docker/Dockerfile [hunk 1]
```diff
@@ -1,8 +1,4 @@
-FROM python:3.8.7-slim
-
-# Update
-RUN apt update -y && apt upgrade -y
-RUN apt install patch gcc libffi-dev wget -y
+FROM kprasch/rust-python:3.8.9
```
Comment (cygnusv) [inline deploy/docker/Dockerfile:1]: Can we host this in a nucypher Docker repo?

File: requirements.txt [hunk 2]
```diff
@@ -104,3 +104,4 @@ web3==5.12.3
 websockets==8.1; python_full_version >= '3.6.1'
 werkzeug==2.0.1; python_version >= '3.6'
 zope.interface==5.4.0; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3, 3.4'
+nucypher_core @ git+https://github.com/nucypher/nucypher-core.git#subdirectory=nucypher-core-python
```
Comment (cygnusv) [inline requirements.txt:107]: Awaiting nucypher-core 0.0.1 on Pypi

**Review by cygnusv — APPROVED** [review]
Thanks @fjarri !

File: requirements.txt (hunk 2 above)
Comment (cygnusv) [reply requirements.txt:107]: Opened https://github.com/nucypher/nucypher-core/issues/3

File: deploy/docker/Dockerfile (hunk 1 above)
Comment (KPrasch) [reply deploy/docker/Dockerfile:1]: Yes.  I'll push those containers to the nucypher docker repo and make a follow-up PR.

**Review by derekpierre — APPROVED** [review]
🎸 . 

Approved the PR, but if there are no objections to the suggestions @piotr-roslaniec and I made, then can someone commit those suggestions as a batch, and then we can merge it. I don't want to commit my own suggestion 😄 .

File: .circleci/docker/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+export DOCKER_USER=<USERNAME>
+
+# the version of python to build
+export PYTHON_VERSION=<VERSION>
+```
+
+### Build
+```bash
+# pass local env vars as build args USER and VERSION
+docker build -t nucypher/rust-python:$PYTHON_VERSION . \
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply .circleci/docker/README.md]: Yes, I'll move these containers over to the nucypher repo first.

File: requirements.txt (hunk 2 above)
Comment (KPrasch) [reply requirements.txt:107]: Done https://pypi.org/project/nucypher-core/

Comment (KPrasch) [reply requirements.txt:107]:
```suggestions
nucypher-core==0.0.1;  python_version >= '3.7'
```

**Review by KPrasch — APPROVED** [review]

**Pull request merged**